pkg/pillar/cmd/volumemgr/handleblob.go

```diff
@@ -0,0 +1,127 @@
+		blobStatus := blobStatusInt.(types.BlobStatus)
+		blobShaAndBlobStatus[blobSha] = &blobStatus
+	}
+	return blobShaAndBlobStatus
+}
+
+func publishBlobStatus(ctx *volumemgrContext, status *types.BlobStatus) {
+	log.Infof("publishBlobStatus(%s)", status.Key())
+
+	pub := ctx.pubBlobStatus
+	key := status.Key()
+	pub.Publish(key, *status)
+}
+
+func unpublishBlobStatus(ctx *volumemgrContext, status *types.BlobStatus) {
+
+	log.Infof("publishPersistImageStatus(%s)", status.Key())
+
+	pub := ctx.pubBlobStatus
+	key := status.Key()
+	st, _ := pub.Get(key)
+	if st == nil {
+		log.Errorf("unpublishBlobStatus(%s) not found", key)
+		return
+	}
+	pub.Unpublish(key)
+}
+
+//casBlobMonitor checks for blob updates periodically
+func casBlobMonitor(ctx *volumemgrContext) {
+	ticker := time.NewTicker(60 * time.Second)
+	for {
+		select {
+		case <-ticker.C:
+			if err := fetchAndPublishBlobs(ctx); err != nil {
+				log.Errorf("casBlobMonitor: exception while fetchAndPublishBlobs: %s", err.Error())
+			}
+		}
+	}
+}
+
+//fetchAndPublishBlobs creates, updates and updates/un-publish blobStates based on the blob available in CAS.
+func fetchAndPublishBlobs(ctx *volumemgrContext) error {
+	changed := false
+	blobList, err := casClient.ListBlobs()
+	if err != nil {
+		err = fmt.Errorf("fetchAndPublishBlobs: exception while fetch blob list from CAS: %s", err.Error())
+		log.Error(err.Error())
+		return err
+	}
+
+	//Handle blob creation and update in CAS
+	for _, blobSha := range blobList {
+		blobSize, err := casClient.GetBlobSize(blobSha)
+		if err != nil {
+			err = fmt.Errorf("fetchAndPublishBlobs: exception while fetch blob %s size from CAS: %s", blobSha, err.Error())
+			log.Error(err.Error())
+			return err
+		}
+		blobStatus := lookupBlobPersistImageStatus(ctx, blobSha)
+		if blobStatus == nil {
+			// Blob present in CAS, but not found in blobStatus (pub)  list.
+			// Creating a new blobStatus for that blob and publishing it.
+			log.Infof("fetchAndPublishBlobs: creating new blobStatus: %s", blobSha)
+			blobStatus = &types.BlobStatus{
+				BlobSha:            blobSha,
+				Size:               blobSize,
+				State:              types.DOWNLOADED,
+				RefCount:           0,
+				ProgressPercentage: 100,
+				LastUse:            time.Now(),
+				Expired:            false,
+			}
+			changed = true
+		} else {
+			if blobStatus.Size != blobSize {
+				log.Infof("fetchAndPublishBlobs: updating blobStatus %s size to: %v", blobSha, blobSize)
+				blobStatus.Size = blobSize
+				changed = true
+			}
+		}
+		if changed {
+			publishBlobStatus(ctx, blobStatus)
+		}
+	}
+
+	//Handle blob deletion in CAS
+	blobShaAndBlobStatus := blobStatusGetAll(ctx) // All known blob.
+	//We have to check if all known blob are present in CAS.
+	// If not, then we should unpublish the blobStatus of the blob which is not found in CAS
+	for blobSha, blobStatus := range blobShaAndBlobStatus {
+		if !casClient.CheckBlobExists(blobSha) {
```

Comment:
Why do you need to call into casClient for this? You have the list of blobs in blobList.

Also, you don't need to pay the overhead of creating a map in blobStatusGetAll.
Instead of can loop over key := range pub.GetAll() , and check if that key/sha exists in blobList. If not it should be unpublished.

Reply:
Changed.